src/psyclone/tests/psyir/nodes/call_test.py

```diff
@@ -628,6 +630,560 @@ def test_call_get_callees_local(fortran_reader):
+
+    routine_foo_c: Routine = root_node.walk(Routine)[3]
+    assert routine_foo_c.name == "foo_c"
+
+    call_foo_c: Call = routine_main.walk(Call)[6]
+    assert call_foo_c.routine.name == "foo"
+
+    (result, arg_idx_list) = call_foo_c.get_callee()
+    result: Routine
+
+    assert len(arg_idx_list) == 2
+    assert arg_idx_list[0] == 1
+    assert arg_idx_list[1] == 0
+
+    assert result is routine_foo_c
+
+
+def test_call_get_callee_6_interfaces_2_2(fortran_reader):
+    '''
+    Check that optional and named arguments have been correlated correctly
+    '''
+
+    root_node: Node = fortran_reader.psyir_from_source(_code_test_get_callee_6)
+
+    routine_main: Routine = root_node.walk(Routine)[0]
+    assert routine_main.name == "main"
+
+    routine_foo_c: Routine = root_node.walk(Routine)[3]
+    assert routine_foo_c.name == "foo_c"
+
+    call_foo_c: Call = routine_main.walk(Call)[7]
+    assert call_foo_c.routine.name == "foo"
+
+    (result, arg_idx_list) = call_foo_c.get_callee()
+    result: Routine
+
+    assert len(arg_idx_list) == 3
+    assert arg_idx_list[0] == 1
+    assert arg_idx_list[1] == 0
+    assert arg_idx_list[2] == 2
+
+    assert result is routine_foo_c
+
+
+def test_call_get_callee_7_matching_arguments_not_found(fortran_reader):
+    '''
+    Trigger error that matching arguments were not found
+    '''
+    code = '''
+module some_mod
+  implicit none
+contains
+
+  subroutine main()
+    integer :: e, f, g
+    ! Use named argument 'd', which doesn't exist
+    ! to trigger an error when searching for the matching routine.
+    call foo(e, f, d=g)
+  end subroutine
+
+  ! Matching routine
+  subroutine foo(a, b, c)
+    integer :: a, b, c
+  end subroutine
+
+end module some_mod'''
+
+    psyir = fortran_reader.psyir_from_source(code)
+
+    routine_main: Routine = psyir.walk(Routine)[0]
+    assert routine_main.name == "main"
+
+    call_foo: Call = routine_main.walk(Call)[0]
+
+    with pytest.raises(CallMatchingArgumentsNotFound) as err:
+        call_foo.get_callee()
+
+    assert "No matching routine found for 'call foo(e, f, d=g)" in str(
+        err.value
+    )
+
+
+def test_call_get_callee_8_arguments_not_handled(fortran_reader):
+    '''
+    Trigger error that matching arguments were not found
+    '''
+    code = '''
+module some_mod
+  implicit none
+contains
+
+  subroutine main()
+    integer :: e, f
+    ! Use name 'd' which doesn't exist
+    call foo(e, f)
+  end subroutine
+
+  ! Matching routine
+  subroutine foo(a, b, c)
+    integer :: a, b, c
+  end subroutine
+
+end module some_mod'''
+
+    psyir = fortran_reader.psyir_from_source(code)
+
+    routine_main: Routine = psyir.walk(Routine)[0]
+    assert routine_main.name == "main"
+
+    call_foo: Call = routine_main.walk(Call)[0]
+
+    with pytest.raises(CallMatchingArgumentsNotFound) as err:
+        call_foo.get_callee()
+
+    assert "No matching routine found for 'call foo(e, f)" in str(err.value)
+
```

Comment:
Please could you also add a test to demonstrate that we don't match a routine if the *types* of the arguments don't match. You'll need to do this for both a named and a positional argument.

Reply:
There's no difference between named and positional. Once the pairs or arguments (from call + routine) have been determined, the types are matched separately to this.

Reply:
Generally speaking, it's best to attempt to test as though you don't know how it is implemented. However, I think the range of tests you have now covers this so I'm happy.